### ISSUE TITLE
Question about loss computation in sparse pretraining

### ISSUE THREAD

**Issue by danielm09** [vishalned/MMEarth-train#7]
Hi.

I am pretraining an adapted version of your code, which basically uses sentinel-2 images only.

I'm using images with size 56 and patch size 8.

I was investigating potential reasons why I was getting a very low loss (< 1e-5) and no apparent difference in the training chart of the finetuning and training from scratch with no pretraining.

I'm not sure if I made some mistake when adapting the code or if it is supposed to be that way, but when I analyzed the `target` tensor used to calculate the mse loss, I saw that the tensor had only zeros in the masked patches (patches where mask == 1).

https://github.com/vishalned/MMEarth-train/blob/f11564b490e8862fb72b12d1b9e29a5b8088e614/models/fcmae.py#L375

Therefore, when you compute the squared error you are comparing the predictions with zeros instead of the actual image values in those patches.
https://github.com/vishalned/MMEarth-train/blob/f11564b490e8862fb72b12d1b9e29a5b8088e614/models/fcmae.py#L384

Upon further investigation, I realized that `imgs` is modified (it is filled with zeros in the masked patches) after going through the encoder, even though no actual modification is made to this variable. I check this behavior with a toy example using the original ConvNextV2 code and confirmed that `imgs` remain the same after the encoder - and hence `target` is a dense tensor (no zeros anywhere).

I guess one workaround is setting this line
https://github.com/vishalned/MMEarth-train/blob/f11564b490e8862fb72b12d1b9e29a5b8088e614/models/fcmae.py#L442
to

`x, mask = self.forward_encoder(imgs.clone(), mask_ratio)`

which apparently prevents `imgs` from being modified, hence making `target` a dense tensor as it should be.

I am not sure if my assessment of this situation is correct, so any insight is appreciated.

**Comment by vishalned** [comment]
Hey, thanks, I will have to check this properly again. In the code, we create a dictionary of different modalities and only reference imgs as `imgs['sentinel2']`. I assumed that this would create a clone and would not alter the dictionary itself. I would have to check this manually to see if what you say is happening. I will do that and get back soon.

**Comment by vishalned** [comment]
Closing since this issue doesn't occur when working with dictionaries. Although when NOT using dictionaries the above solution works.
